### ISSUE TITLE
Build bots fail on dynamic compiler format ever since max length self-tests were introduced.

### ISSUE THREAD

**Issue by magnumripper** [magnumripper/JohnTheRipper#1751]
Since f545cabcd

I could understand if it was just CircleCI (non SIMD build) but I can't see why Travis sees problems I do not.

Travis
```
Testing: dynamic=md5($p) [128/128 XOP 4x2]... FAILED (cmp_all(4))
```

CircleCI
```
Testing: dynamic=md5($p) 32/64 x2 (MD5_body) [dynamic_6000]... FAILED (cmp_all(2))
```

**Comment by magnumripper** [comment]
BTW @jfoug there's something odd with the algo name for non-SIMD (CircleCI above). It says
```
Testing: dynamic=md5($p) 32/64 x2 (MD5_body) [dynamic_6000]
```
It should say
```
Testing: dynamic=md5($p) [32/64 x2 (MD5_body)]
```

**Comment by jfoug** [comment]
Dynamic is actually running dynamic_6000 (internally).  That is just how the self-describing dyna stuff did it.  It may actually be other numbers (6001, 6002, etc), depending upon how many self-describing we do. If we had a whole dyna.conf full of them, then each time john ran, it would build the ones needed.   I guess I could mute it's output, but when I wrote it, I did so in a way that dyna would just work, it has no idea that the script generated came from a transient script, built on the fly.

Look at any expr,debug output. You will see there is a dynamic number in there.  The number means nothing. There will never be a number output to the .pot file. also _6xxxx is a range that is not valid to have a script in anyway.

**Comment by magnumripper** [comment]
Sure but the problem is you are putting the wrong stuff in algorithm_name. It only happens on non-SIMD.

**Comment by magnumripper** [comment]
Look at my first post in this issue. The XOP build has it right.

**Comment by magnumripper** [comment]
Oh, I just found how to reproduce. It doesn't fail on its own!
```
$ ../run/john -test=0 -form:dynamic='md5($p)'
Testing: dynamic=md5($p) [128/128 AVX 4x3]... PASS

$ ../run/john -test=0 -form:dynamic
Testing: dynamic_0 [md5($p) (raw-md5) 128/128 AVX 4x3]... PASS
Testing: dynamic_1 [md5($p.$s) (joomla) 128/128 AVX 4x3]... PASS
Testing: dynamic_2 [md5(md5($p)) (e107) 128/128 AVX 4x3]... PASS
(...)
Testing: asa-md5, Cisco ASA [md5($p.$s) (Cisco ASA) 128/128 AVX 4x3]... PASS
Testing: dynamic=md5($p) [128/128 AVX 4x3]... FAILED (cmp_all(4))

Testing: FormSpring [sha256($s.$p) 128/128 AVX 4x]... PASS
Testing: hMailServer [sha256($s.$p) 128/128 AVX 4x]... PASS
Testing: MediaWiki [md5($s.md5($p)) 128/128 AVX 4x3]... PASS
Testing: md5ns, Netscreen [md5($s.$p) (OSC) (PW > 31 bytes) 128/128 AVX 4x3]... PASS
Testing: osc, osCommerce [md5($s.$p) (OSC) 128/128 AVX 4x3]... PASS
Testing: PHPS [md5(md5($p).$s) 128/128 AVX 4x3]... PASS
Testing: PHPS2 [md5(md5($p).$s) 128/128 AVX 4x3]... PASS
Testing: pix-md5, Cisco PIX [md5($p) (Cisco PIX) 128/128 AVX 4x3]... PASS
1 out of 140 tests have FAILED
```

**Comment by magnumripper** [comment]
@jfoug 

The problem is I'm using `pPriv->pSetup->MaxInputLen` to know my current max length in dynamic_compiler_fmt_plug.c, link_funcs().

Run alone it works fine:
```
$ ../run/john -test=0 -form:dynamic='md5($p)'
our_init()
link_funcs()
maxlen 55
Testing: dynamic=md5($p) [128/128 AVX 4x3]... PASS
```

But when not alone, it's zero!
```
$ ../run/john -test=0 -form:dynamic
Testing: dynamic_0 [md5($p) (raw-md5) 128/128 AVX 4x3]... PASS
Testing: dynamic_1 [md5($p.$s) (joomla) 128/128 AVX 4x3]... PASS
(...)
Testing: asa-md5, Cisco ASA [md5($p.$s) (Cisco ASA) 128/128 AVX 4x3]... PASS
our_init()
link_funcs()
maxlen 0
Testing: dynamic=md5($p) [128/128 AVX 4x3]... FAILED (cmp_all(4))
```

**Comment by magnumripper** [comment]
OK, using `fmt_CompiledDynamic.params.plaintext_length` might be better. But it's not only that. The problem is also (I think) that I create the max-length test vector prior to optimizations, and the max length may shrink...

**Comment by magnumripper** [comment]
OK, I simply needed to supply a test vector for length 55 in the optimized version in dynamic_compiler_lib.c... I only did so for `md5($p)` for now. We'll see what the bots say.

**Comment by magnumripper** [comment]
Ah, finally Travis completes without problem. Now only the CircleCI non-SIMD build fail.

**Comment by magnumripper** [comment]
Fixed in 2948895. I'm not very happy with the hacks but it should work.

**Comment by jfoug** [comment]
Ok, I do have the TestM= and TestF= in the Dyna_parser language (the part that takes a script and generates the format.  I wonder if something like that would 'help'.   I know that I have split out the hash from the test string, but possibly, I could change that code.  That way we could have the 55 byte hash using the TestM, and the 110 byte with the TestF  (I think that is the order).

Might be something to think about.  Your logic is not that bad.  I have had to do things like that numerous times in dyna in other places.  It happens.
